google-cloud-pubsub/lib/google/cloud/pubsub/subscriber.rb

```diff
@@ -0,0 +1,217 @@
+# Copyright 2017 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+require "google/cloud/pubsub/service"
+require "google/cloud/errors"
+require "monitor"
+require "forwardable"
+require "concurrent"
+
+module Google
+  module Cloud
+    module Pubsub
+      ##
+      # # Subscriber
+      #
+      class Subscriber
+        include MonitorMixin
+
+        ##
+        # @private Implementation attributes.
+        attr_reader :request_queue, :output_enum, :thread_pool, :service
+
+        ##
+        # Subscriber attributes.
+        attr_reader :callback, :subscription, :deadline, :threads
+
+        ##
+        # @private Create an empty {Subscriber} object.
+        def initialize callback, subscription, deadline, threads, service
+          @request_queue = nil
+          @output_enum = nil
+          @callback = callback
+          @subscription = subscription
+          @deadline = deadline
+          @threads = threads || [2, Concurrent.processor_count * 2].max
+          @service = service
+
+          super() # to init MonitorMixin
+        end
+
+        def start
+          synchronize do
+            return if @request_queue
+
+            @request_queue = EnumeratorQueue.new(self)
+            @thread_pool = Concurrent::FixedThreadPool.new(
+              threads,
+              fallback_policy: :caller_runs
+            )
+            start_streaming!
+          end
+
+          true
+        end
+
+        def stop
+          synchronize do
+            return if @request_queue.nil?
+            @request_queue.push self
+          end
+
+          true
+        end
+
+        def wait!
```

Comment:
We haven't committed time to the documentation of these methods since we aren't sure they are correct. But I will do my best to add some dev document them here:

Subscriber#stop only halts the pulling of new messages. It does not ensure that all messages have been processed. It updates the state of the Subscriber object and will return quickly.

Subscriber#wait! will block until all the current messages in the queue are processed. Because it blocks it has the bang to indicate that it could take a while. This is similar to BigQuery's `Job#wait_until_done!`, Spanner's `Job#wait_until_done!`, Speech's `Operation#wait_until_done!`, Speech's `Stream#wait_until_complete!`, and DNS's `Change#wait_until_done!`. In all of these other cases these object will complete eventually, but in Pub/Sub the Subscription is intended to operate open ended. There is no logical end for how long it will run, it will only be stopped eventually.

If it were me, I'd name `wait!` something like `wait_until_empty!`.